python/ray/train/_internal/checkpoint_manager.py

```diff
@@ -0,0 +1,218 @@
+    ):
+        self._checkpoint_config = checkpoint_config or CheckpointConfig()
+
+        # Incremental unique checkpoint ID of this run.
+        self._latest_checkpoint_index = latest_checkpoint_index
+
+        # A heap that tracks the top K checkpoints.
+        self._top_k_checkpoints: List[_HeapCheckpointWrapper] = []
+
+        # The most recently registered checkpoint.
+        # The latest checkpoint should never be deleted immediately,
+        # even if it's the worst checkpoint.
+        self._latest_checkpoint: Optional[_TrackedCheckpoint] = None
+
+        # The best checkpoint seen so far.
+        # Since _top_k_checkpoints is a min-heap (to know who to evict),
+        # we still need to keep track of the best checkpoint independently.
+        self._best_checkpoint: Optional[_HeapCheckpointWrapper] = None
+
+        # Deletion of some checkpoints should be deferred. Specifically, if the
+        # latest persisted checkpoint should be deleted, we will only delete it
+        # once a new checkpoint came in (so that `_latest_checkpoint` is available).
+        self._checkpoints_to_clean_up: Set[_TrackedCheckpoint] = set()
+
+    def register_checkpoint(self, tracked_checkpoint: _TrackedCheckpoint):
+        """Register new checkpoint and add to bookkeeping.
+
+        This method will register a new checkpoint and add it to the internal
+        bookkeeping logic. This means the checkpoint manager will decide if
+        this checkpoint should be kept, and if older or worse performing
+        checkpoints should be deleted.
+
+        Args:
+            checkpoints: Tracked checkpoint object to add to bookkeeping.
+        """
+        # Mark the index of the checkpoint, for use in determining the
+        # order in which checkpoints were registered.
+        # TODO(justinvyu): Should this be unified with the checkpoint index
+        # tracked in the session / used in the checkpoint dirname?
+        tracked_checkpoint.index = self._latest_checkpoint_index
+        self._latest_checkpoint = tracked_checkpoint
+
+        # Add to heap.
+        checkpoint_priority = self._get_checkpoint_score(tracked_checkpoint)
+        wrapped_checkpoint = _HeapCheckpointWrapper(
+            checkpoint_priority, tracked_checkpoint
+        )
+        heapq.heappush(self._top_k_checkpoints, wrapped_checkpoint)
+
+        # Update best checkpoint.
+        if (
+            self._best_checkpoint is None
+            or wrapped_checkpoint.priority >= self._best_checkpoint.priority
+        ):
+            self._best_checkpoint = wrapped_checkpoint
+
+        # Delete the worst checkpoint if we're over the `num_to_keep` limit.
+        if (
+            self._checkpoint_config.num_to_keep is not None
+            and len(self._top_k_checkpoints) > self._checkpoint_config.num_to_keep
+        ):
+            worst_checkpoint = heapq.heappop(self._top_k_checkpoints)
+            self._maybe_delete_checkpoint(worst_checkpoint.tracked_checkpoint)
```

Comment:
Why not cleanup checkpoints right away instead of deferring?

Reply:
This is behavior from before: we always keep the latest checkpoint around, even if it's not a top k checkpoint. Restoration from fault tolerance restores from the latest checkpoint at the moment.

Reply:
@ericl Are you ok with keeping the existing functionality for now? [This test walks through an example.](https://github.com/ray-project/ray/pull/37962/files#diff-d2d57b4ec2ce9f48e139622dd4f22bb22f6e19145a7d2e223fcc1fde5e510034R111)

Reply:
Hmm I'm more talking about the implementation. I think now's a good time to clean up some of this code. I believe we can replace this class with the following, which is much simpler and unlikely to be much slower:


```
class _CheckpointManager:

    def __init__(
        self,
        checkpoint_config: Optional[CheckpointConfig],
        latest_checkpoint_index: int = 0,
    ):
        self._checkpoint_config = checkpoint_config or CheckpointConfig()
        self._checkpoints: List[_TrackedCheckpoint)] = []
        self._latest_checkpoint_index = latest_checkpoint_index

    def register_checkpoint(self, tracked_checkpoint: _TrackedCheckpoint):
        self._checkpoints.append(tracked_checkpoint)

        if self._checkpoint_config.num_to_keep:
            # Could cache score as optimization if we wanted
            candidates = sorted(self._checkpoints[:-1], key=self._get_checkpoint_score)
            while candidates and len(self._checkpoints) > self._checkpoint_config.num_to_keep:
                checkpoint = candidates.pop(0).checkpoint
                _delete_fs_path(fs=checkpoint.filesystem, fs_path=checkpoint.path)

    @property
    def best_checkpoint(self) -> Optional[_TrackedCheckpoint]:
        if not self._checkpoints:
            return None
        return sorted(self._checkpoints, key=self._get_checkpoint_score)[-1]

    @property
    def latest_checkpoint(self) -> Optional[_TrackedCheckpoint]:
        if not self._checkpoints:
            return None
        return self._checkpoints[-1]
```

Reply:
Ok, simplified it in a slightly different way that doesn't need to sort the list. PTAL

There was also a slight behavioral change from above: the K+1th checkpoint wouldn't be deleted if the best checkpoint is always excluded from the candidates.